code/API_definitions/qod-api.yaml

```diff
@@ -419,6 +438,17 @@ components:
       required:
         - severity
         - description
+
+    QosStatus:
+      type: string
+      enum:
+        - REQUESTED
+        - AVAILABLE
+        - UNAVAILABLE
+      description: |
```

Comment:
I'm not sure that we have the information (event) to set the status to AVAILABLE. Although if so, we could set AVAILABLE immediately on the request and not use REQUESTED.

Reply:
If the objective of this new QosStatus is to allow an inmediate response to `createSession` without knowing if the requested QoS can be granted (allowing an asynchronous provision of the QoS), I see the `requested` in the context of the response, but not for the notification event. In that case, a subsequent event may carry values for `available` or `unavailable`. 
Also, `unavailable` is not appropriate for the `createSession` response, because in that case an error would make more sense instead of returning this value along with a session id. 
I have doubs about the use of `available` in the response. A 201 response should imply that the request is fulfilled. If not, a 202 response would make more sense. Or return a more generic 200 and rely on the status.

Reply:
@jlurien
I tend to agree with you, however I think 201 is the correct response whether the QoS profile is granted or only requested because in both cases a session is created which can be later queried, modified (if we add PUT or PATCH) or deleted. And given this, the API caller anyway has to check the qosStatus field to determine if the requested QoS has been granted or is still pending.

So it does not change the API caller's workflow to find out that, in fact, the requested QoS could not be granted and they must either accept that or request a different QoS profile. But it is not clear to me in this case if the session persists (can it still be modified or queried when qosStatus is `UNAVAILABLE`?) or the session is immediately and automatically deleted.

I also can't think of a use case where a notification would have qosStatus as `REQUESTED`.

@emil-cheung
When qosStatus is `UNAVAILABLE`, I think it would be useful to give an indication in the `reason` field as to whether this is only temporary (e.g. due to network congestion) or permanent (e.g. because that QoS profile is just not available where the UE is). So maybe values `CONGESTION` and `NOT AVAILABLE IN THIS LOCATION` could be added as valid reasons why the requested QoS profile is not available. This reason would also need to be included in the 201 response when the qosStatus is `UNAVAILABLE`.

Reply:
@eric-murray
I was initially thinking about a session being created once the request is granted, but that would imply having different ids for requests and sessions. To keep things simpler,  we may consider a session lifecycle starting with "requested", and then moving either to "available/active" or to "unavailable/not_granted".  If a provider is able to resolve the request synchronously it could return the result inmediately.

I see a different status for "terminated", for a session that was "active", and that status may be enriched with reasons like time_expired, abnormal_termination, etc. Status "unavailable/not_granted", for session that was "requested" but never got "active", would have a different set of reasons, like the ones you mention.